lm_eval/tasks/qasper.py

```diff
@@ -0,0 +1,222 @@
+                    {
+                        "title": doc["title"],
+                        "abstract": doc["abstract"],
+                        "question": question,
+                        "answer": answer,
+                        "answer_type": answer_type,
+                    }
+                )
+        return obs_list
+
+    def process_results(self, doc, results):
+        # TODO: Calculate a score for extractive spans once a request type for generating
+        # extractive spans is available
+        if len(results) == 1:
+            [(logprob_unanswerable, _)] = results
+        elif len(results) == 2:
+            res, (logprob_unanswerable, _) = results
+        else:
+            ll_yes, ll_no, (logprob_unanswerable, _) = results
+        res_dict = {}
+
+        # Handle unanswerability first
+        unanswerable_gold = doc["answer_type"] == "unanswerable"
+        unanswerable_pred = exp(logprob_unanswerable) > 1 - exp(logprob_unanswerable)
+        res_dict["f1_un"] = (unanswerable_gold, unanswerable_pred)
+
+        # Handle yes/no questions
+        if doc["answer_type"] == "bool":
+            gold = 1 if doc["answer"] == "yes" else 0
+            pred = ll_yes > ll_no
+            res_dict["f1_yn"] = (gold, pred)
+
+        # Handle completions
+        if doc["answer_type"] == "free form answer":
+            res_dict["f1_ab"] = token_f1_score(res, doc["answer"])
+
+        # Handle extraction
+        if doc["answer_type"] == "extractive_spans":
+            res_dict["f1_ex"] = 0
+        return res_dict
+
+    def aggregation(self):
+        return {"f1_un": f1_score, "f1_yn": f1_score, "f1_ab": mean, "f1_ex": mean}
+
+    def construct_requests(self, doc, ctx):
+        """Uses RequestFactory to construct Requests and returns an iterable of
+        Requests which will be sent to the LM.
+
+        :param doc:
+            The document as returned from training_docs, validation_docs, or test_docs.
+        :param ctx: str
+            The context string, generated by fewshot_context. This includes the natural
+            language description, as well as the few shot examples, and the question
+            part of the document for `doc`.
+        """
+        unanswerable = rf.loglikelihood(ctx, " " + "unanswerable")
+        if doc["answer_type"] in ("free form answer", "extractive_spans"):
+            return [rf.greedy_until(ctx, ["\n"]), unanswerable]
+        elif doc["answer_type"] in ("bool"):
+            ll_yes, _ = rf.loglikelihood(ctx, " yes")
+            ll_no, _ = rf.loglikelihood(ctx, " no")
+            return [ll_yes, ll_no, unanswerable]
+        else:
+            return [unanswerable]
+
+    def higher_is_better(self):
+        """
+        :returns: {str: bool}
+            A dictionary where keys are the names of submetrics and values are
+            whether a higher value of the submetric is better
+        """
+        return {"f1_un": True, "f1_yn": True, "f1_ab": True, "f1_ex": True}
```

Comment:
nit: can you make these the full names? like f1_unanswerable, f1_yesno, etc

Reply:
Done